emergence_lib/src/camera.rs

```diff
@@ -186,61 +165,58 @@ fn camera_zoom(
         .cursor_position()
         .map(|cursor_pos| (cursor_pos / window_size) * 2. - Vec2::ONE);
 
-    if cam.enabled {
-        let old_scale = proj.scale;
-        proj.scale = (proj.scale * (1. + -scroll * 0.001)).max(cam.min_scale);
+    let old_scale = proj.scale;
+    proj.scale = (proj.scale * (1. + -scroll * 0.001)).max(cam.min_scale);
 
-        // Apply max scale constraint
-        if let Some(max_scale) = cam.max_scale {
-            proj.scale = proj.scale.min(max_scale);
-        }
+    // Apply max scale constraint
+    if let Some(max_scale) = cam.max_scale {
+        proj.scale = proj.scale.min(max_scale);
+    }
+
+    // If there is both a min and max x boundary, that limits how far we can zoom. Make sure we don't exceed that
+    if let (Some(min_x_bound), Some(max_x_bound)) = (cam.min_x, cam.max_x) {
+        let max_safe_scale = max_scale_within_x_bounds(min_x_bound, max_x_bound, &proj);
+        proj.scale = proj.scale.min(max_safe_scale);
+    }
```

Comment:
What if there's only one of each?

Reply:
Valid. This was @bzm3r's code: I didn't refactor this bit.